### PR TITLE
Add support for index_select OP (backend implementation)

### PR DESCRIPTION
**Pull request by azazhu**
This PR is backend implementation of index_select(lookup_tv, dim, index_tv) fusion: 
- support arbitrarily rank
- support arbitrarily dim value
- index_tv can become consumer;
- lookup_tv cannot become consumer.

OP def is here: https://pytorch.org/docs/stable/generated/torch.index_select.html

TODO: move index calculation from codegen.cpp to lower_index.cpp

### CODE REVIEW

**Comment by naoyam** [comment]
@zasdfgbnm added support of `select` in #2179. I think it'd be relatively straightforward to build `index_select` on top of his work. Please take a look at the PR, especially around here: https://github.com/csarofeen/pytorch/pull/2179/files#diff-3c08060d5a1993c9599d5cbb2d7a805258ae1b4e166682b497c4f5fd376ba5d3R199-R205

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -34,6 +34,37 @@ TensorView* maybe_broadcast_inner_to_rank(TensorView* t, size_t rank) {
   return t;
 }
 
+TensorView* maybe_broadcast_idxtv_inner_outter_to_rank(
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: Why broadcasting is needed? I have a feeling that if we modify https://github.com/csarofeen/pytorch/blob/c9f8c1de5f3bd6ac0a0cce50569410abe2dfc209/torch/csrc/jit/codegen/cuda/root_domain_map.cpp#L65 to correctly handle pairwise root domain mapping, it might not be needed to broadcast here.

File: test/cpp/jit/CMakeLists.txt [hunk 2]
```diff
@@ -111,6 +111,7 @@ if(USE_CUDA)
   list(APPEND JIT_TEST_SRCS ${TORCH_SRC_DIR}/csrc/jit/codegen/cuda/test/test_gpu_transpose.cpp)
   list(APPEND JIT_TEST_SRCS ${TORCH_SRC_DIR}/csrc/jit/codegen/cuda/test/test_gpu_rng.cu)
   list(APPEND JIT_TEST_SRCS ${TORCH_SRC_DIR}/csrc/jit/codegen/cuda/test/test_gpu_utils.cpp)
+  list(APPEND JIT_TEST_SRCS ${TORCH_SRC_DIR}/csrc/jit/codegen/cuda/test/test_index_select.cpp)
```
Comment (zasdfgbnm) [inline test/cpp/jit/CMakeLists.txt]: Could you move the test to `torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp`?
It is a new file added in https://github.com/csarofeen/pytorch/pull/2179/files#diff-ef65e50bd5d7c7c2e6e34f57fa8d01e977252ced3f6e5f77af1cbcdef2cc744a?

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: yes, we could remove broadcast but we have to modify both PairwiseRootDomainMap::map and ComputeAtRootDomainMapBuilder::mapPointwiseOrReductionOp. Or if we keep broadcast, we can reuse mapPointwiseOrReductionOp. I prefer reusing mapPointwiseOrReductionOp func and adding broadcast, see https://github.com/csarofeen/pytorch/pull/2183/files#diff-a4568d75caaa6e7e45ede61dc386af88660f2ab3e63175ba52074b10d7ebc85aR414. Does it make sense?

**Comment by azazhu** [comment]
> @zasdfgbnm added support of `select` in #2179. I think it'd be relatively straightforward to build `index_select` on top of his work. Please take a look at the PR, especially around here: https://github.com/csarofeen/pytorch/pull/2179/files#diff-3c08060d5a1993c9599d5cbb2d7a805258ae1b4e166682b497c4f5fd376ba5d3R199-R205

Referred to Xiang's impl and moved lower_index from codegen.cpp to lower_index.cpp

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I think I prefer adapting the root domain maps for `index_select`. Unless strictly required, we should not add extra expressions, especially broadcasts.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 3]
```diff
@@ -93,6 +93,11 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseRootDomainMap::map(
   if (SelectOp* sop = dynamic_cast<SelectOp*>(consumer_tv_->definition())) {
     selected_id = sop->getSelectAxis();
   }
+  IterDomain* idx_selected_id = nullptr;
+  if (IndexSelectOp* sop =
+          dynamic_cast<IndexSelectOp*>(consumer_tv_->definition())) {
+    idx_selected_id = sop->getSelectAxis();
+  }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: If we see this pattern often, we should consider adding some routines to handle both `SelectOp` and `IndexSelectOp` uniformly.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: @naoyam @csarofeen  What do you think? It looks like `index_select` is just a broadcast of gather:

```python
import torch

def my_index_select(self, dim, index):
    broadcast_shape = [1] * len(self.shape)
    broadcast_shape[dim] = index.shape[0]
    expand_shape = list(self.shape)
    expand_shape[dim] = index.shape[0]
    index = index.view(broadcast_shape).expand(expand_shape)
    return self.gather(dim, index)

x = torch.randn(5, 6, 7)
index = torch.tensor([1, 2, 3])

assert torch.eq(x.index_select(0, index), my_index_select(x, 0, index)).all()
assert torch.eq(x.index_select(1, index), my_index_select(x, 1, index)).all()
assert torch.eq(x.index_select(2, index), my_index_select(x, 2, index)).all()
```

And if we add broadcasting here, this PR effectively implements `gather` (which is good I guess). In the future gather PR, we might just rename the op name from `IndexSelectOp` to `GatherOp`.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp [hunk 4]
```diff
@@ -131,6 +131,642 @@ TEST_F(NVFuserTest, FusionSelectOpPersistent_CUDA) {
...
+
+  TensorView* tv_sel = index_select(tv0, 0, tv_idx);
+  // Register your outputs
+  fusion.addOutput(tv_sel);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor input0 = at::randn({nElem, nFeat}, options); // lookup
+  std::vector<int64_t> storage(nElem_select);
+  for (int i = 0; i < nElem_select; ++i) {
+    storage[i] = std::rand() % nElem;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp]: Let's make this deterministic. Using rand is fine, but this means selected indices are non deterministic, which can make debugging harder. Please deterministically initialize a random seed.

It may also be a good idea to have multiple runs with different index values, but they should be deterministically generated. So, for example:

```
for (int i = 0; i < 10; ++i) {
   int rand_seed = i;
   // initialize rand with rand_seed
   std::vector<int64_t> storage(nElem_select);
   for (int i = 0; i < nElem_select; ++i) {
     storage[i] = std::rand() % nElem;
      ...
}
```

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 5]
```diff
@@ -196,6 +196,19 @@ void IndexLowering::handle(const TernaryOp* top) {
   GpuLower::current()->propagateExprInfo(top, back());
 }
 
+void IndexLowering::handle(const IndexSelectOp* sop) {
+  const auto indices = lowerSrcIndex(sop->input(1), sop->output(0));
+  const std::unordered_map<IterDomain*, Val*> override_index = {
+      {sop->getSelectAxis(), indices}};
+  const auto lookup =
+      lowerSrcIndex(sop->input(0), sop->output(0), override_index);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp:204]: This is surprisingly clean!

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: IIUC, `gather` is more flexible than `index_select`, but at this point since we assume inputs are fusion inputs, the added flexibility doesn't matter. Since we would eventually need to lift the assumption, don't we want to have them as separate operations?

File: torch/csrc/jit/codegen/cuda/transform_replay.cpp [hunk 6]
```diff
@@ -286,7 +286,8 @@ std::pair<TensorDomain*, unsigned int> TransformReplay::replayPasC(
     auto it = replay_PasC.getReplay().find(c_id);
     if (it == replay_PasC.getReplay().end()) {
       TORCH_INTERNAL_ASSERT(
-          c_id->isBroadcast() || c_id->isGather() || c_id->isVectorComponent(),
+          c_id->isBroadcast() || c_id->isGather() ||
+              c_id->isVectorComponent() || is_producer_lookup_tv,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_replay.cpp]: Allowing any `c_id` when the producer is an input to `index_select` doesn't seem right. We should check if `c_id` depends on the selected dimension. 

You could do something like:

```
IterDomain* consumer_root_selected_domain = ...;
bool c_id_depends_on_selectd_domain = DependencyCheck::isDependencyOf(consumer_root_selected_domain, c_id);
...
TORCH_INTERNAL_ASSERT(
          c_id->isBroadcast() || c_id->isGather() ||
              c_id->isVectorComponent() || c_id_depends_on_selectd_domain,
```

**Comment by naoyam** [comment]
@azazhu Can you please also add a validation check to `validateIr` so that all of the lookup TVs are ensured to be fusion inputs? Something like:

```
for (auto expr: fusion->exprs()) {
   if (expr->isA<SelectOp>() || expr->isA<IndexSelectOp>()) {
     TORCH_CHECK(expr->input(0)->isFusionInput(), "Lookup input must be a fusion input: ", expr->toString());
   }
}
```

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: > IIUC, `gather` is more flexible than `index_select`, but at this point since we assume inputs are fusion inputs, the added flexibility doesn't matter. Since we would eventually need to lift the assumption, don't we want to have them as separate operations?

This is a good point. I agree.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp (hunk 4 above)
Comment (chang-l) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp]: I wonder here why not simply using `at::randint()` to generate random indexes?

File: torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp [hunk 7]
```diff
@@ -131,6 +131,642 @@ TEST_F(NVFuserTest, FusionSelectOpPersistent_CUDA) {
...
+  }
+  auto opts = torch::TensorOptions().dtype(torch::kLong);
+  auto input_idx_cpu =
+      torch::from_blob(storage.data(), {int64_t(storage.size())}, opts).clone();
+  auto input_idx = input_idx_cpu.to(torch::kCUDA);
+  at::Tensor output =
+      at::randn({nElem_select, nFeat}, options); // output&elemwise
```
Comment (chang-l) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp]: here, instead, maybe use `auto output = at::empty({dim1, dim2}, options)`?

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 3 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: For this case, SelectOp and IndexSelectOp have different behaviors, see https://github.com/csarofeen/pytorch/pull/2183/files#diff-147f701ca808989bf9ad700751c261c21fc30cea6c0f434001cdc4c2ce0be042R113-R119. For rest cases, like https://github.com/csarofeen/pytorch/pull/2183/files#diff-13bc8a76736a772a30a4cf860ec9527385f8c35d4c8481d9eeff654e6f5f8a5bR1225-R1252 , we should cosider adding some routines. Right now, I added a TODO. How about refining them once gather/scatter are added?

File: torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp (hunk 4 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp]: Done, added seed into all UTs and added multiple runs on some UTs.

File: torch/csrc/jit/codegen/cuda/transform_replay.cpp (hunk 6 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/transform_replay.cpp]: you're right. Fixed, see https://github.com/csarofeen/pytorch/pull/2183/files#diff-3051147c9c0739bc571fb8109b8333e682245de4b0aaf8e5e18d6a3419c43c51R287-R294

**Comment by azazhu** [comment]
> @azazhu Can you please also add a validation check to `validateIr` so that all of the lookup TVs are ensured to be fusion inputs? Something like:
> 
> ```
> for (auto expr: fusion->exprs()) {
>    if (expr->isA<SelectOp>() || expr->isA<IndexSelectOp>()) {
>      TORCH_CHECK(expr->input(0)->isFusionInput(), "Lookup input must be a fusion input: ", expr->toString());
>    }
> }
> ```

Done. see https://github.com/csarofeen/pytorch/pull/2183/files#diff-e9719a5d2b00c30404cd4d11ebaf37dea60b9447ee893aaddb9a9ef6fd9aaebbR1313-R1323

File: torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp (hunk 4 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp]: Done. Replaced by at::randint().

File: torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp (hunk 7 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_indexing_ops.cpp]: Good caught. Done. replaced by at::zeros.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: Xiang, I removed this broadcast in local branch(looks ok), but this issue https://github.com/csarofeen/pytorch/issues/2216 blocks me. Could we just keep the broadcast in this PR and then I will create another PR for index_select(removing broadcast) after https://github.com/csarofeen/pytorch/issues/2216 fixed?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: @naoyam Without broadcasting, we do have issues with transform replay. For example
```
T0[I0] ...
T1[I1, I2, I3] = input
T2[I1, I0, I3] = index_select(lookup_tv=T1, dim=1, index_tv=T0)
```
and if `T2` is replayed as `[(I1*I0*I3)/128, 128]`, what should `T0` be replayed as? Do we want `[I0/128, 128]`? Or a more general question is, do we want to enable the forwarding for all unmapped IterDomains in the consumer, instead of just broadcasting?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I see. Looks like broadcasting the index tensor seems reasonable then. Alternatively, we could extend the transform forwarding for index_select, but I don't know which would be better. Broadcasting seems most straightforward for index_select, so I think that would make most sense.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: Yeah, that's fine.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: Yeah, let's keep the broadcast then.

Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: typo: outer

Actually, just `maybe_broadcast_index_tv` sounds good too.

**Review by naoyam — APPROVED** [review]
LGTM. Just left one more minor comment. 

Thanks for adding this new operator!

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: done
